### PR TITLE
Climber testing

### PR DESCRIPTION
**Pull request by yutytuty**

### CODE REVIEW

File: src/main/java/frc/robot/Subsystems/ClimberWinch.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+import frc.robot.RobotMap;
+
+import java.util.function.Supplier;
+
+/**
+ * Controls the climber winch which controls the height of the telescopic arms.
+ */
+public class ClimberWinch extends MotoredGenericSubsystem {
+
+    private Supplier<Double> DOWN_SPEED = rootNamespace.addConstantDouble("down speed", -0.25);
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/Subsystems/ClimberWinch.java:20]: why UPPER_SNAKE_CASE?

File: src/main/java/frc/robot/Subsystems/ClimberWinch.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    private Supplier<Double> HOOKED_DOWN_SPEED = rootNamespace.addConstantDouble("hooked down speed", -0.6);
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/Subsystems/ClimberWinch.java:21]: ditto

File: src/main/java/frc/robot/Subsystems/ClimberWinch.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    private Supplier<Double> UP_SPEED = rootNamespace.addConstantDouble("up speed", 0.4);
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/Subsystems/ClimberWinch.java:22]: ditto

**Review by OfriRosenbaum — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/commands/DropPlacer.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+import frc.robot.subsystems.ClimberPlacer;
+
+import java.util.function.Supplier;
+
+public class DropPlacer extends MoveGenericSubsystem {
+
+    private final ClimberPlacer placer;
+
+    private static final RootNamespace rootNamespace = new RootNamespace("drop placers");
+    public static final Supplier<Double> initialWaitTime = rootNamespace.addConstantDouble("initial wait time", 0.25);
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/commands/DropPlacer.java:15]: convention A-I\a2

File: src/main/java/frc/robot/RobotMap.java [hunk 5]
```diff
@@ -21,6 +21,9 @@
 
         int CLIMBER_WINCH_SPARK_MAX_1 = 12;
         int CLIMBER_WINCH_SPARK_MAX_2 = 13;
+
+        int CLIMBER_PLACER_TALON_LEFT = -1;
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/RobotMap.java:25]: Add TODO of changing port (to this, line 26, lines 43-46).
You don't have to if you remember to change it, but if you forget I'll make fun of you for not listening to me

File: src/main/java/frc/robot/subsystems/ClimberPlacer.java [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+        this.backLimit = new DigitalInput(backLimitPort);
+    }
+
+    @Override
+    public boolean canMove(double speed) {
+        return !(frontLimit.get() && speed < 0) && !(backLimit.get() && speed > 0);
+    }
+
+    @Override
+    public void configureDashboard() {
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/subsystems/ClimberPlacer.java:58]: I'd add the option to drop the placers individually to potentially help with testing

**Review by Yoel-Perman-Brilliant — CHANGES_REQUESTED** [review]
looks good, just add OI

File: src/main/java/frc/robot/commands/DropPlacer.java (hunk 4 above)
Comment (yutytuty) [reply src/main/java/frc/robot/commands/DropPlacer.java:15]: Convention Yotam-a12345678: write the convention in the comment so I don't need to look it up in the file.

File: src/main/java/frc/robot/commands/MoveBetweenBars.java [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+ * Moves the dynamic climbers downward towards the ground until they hit the third bar.
+ */
+public class MoveBetweenBars extends SequentialCommandGroup {
+
+    public MoveBetweenBars() {
+        ClimberWinch climberWinch = ClimberWinch.getInstance();
+        addCommands(
+                new MoveGenericSubsystem(climberWinch, ClimberWinch.UP_SPEED),
+                new DropBothPlacers(),
+                new MoveGenericSubsystem(ClimberWinch.getInstance(), ClimberWinch.DOWN_SPEED)
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/commands/MoveBetweenBars.java:17]: use the instance

File: src/main/java/frc/robot/commands/MoveBetweenBars.java [hunk 8]
```diff
@@ -0,0 +1,20 @@
+package frc.robot.commands;
+
+import com.spikes2212.command.genericsubsystem.commands.MoveGenericSubsystem;
+import edu.wpi.first.wpilibj2.command.SequentialCommandGroup;
+import frc.robot.subsystems.ClimberWinch;
+
+/**
+ * Moves the dynamic climbers downward towards the ground until they hit the third bar.
+ */
+public class MoveBetweenBars extends SequentialCommandGroup {
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/commands/MoveBetweenBars.java:10]: Wow, this is a great piece of code! Perhaps, you should use it in your code. At least that's what I would do if I were you.

**Review by OfriRosenbaum — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/commands/MoveBetweenBars.java (hunk 8 above)
Comment (Yoel-Perman-Brilliant) [reply src/main/java/frc/robot/commands/MoveBetweenBars.java]: damnnnnn

File: src/main/java/frc/robot/OI.java [hunk 9]
```diff
@@ -4,6 +4,7 @@
 import com.spikes2212.util.XboxControllerWrapper;
 import edu.wpi.first.wpilibj.Joystick;
 import edu.wpi.first.wpilibj2.command.ParallelCommandGroup;
+import edu.wpi.first.wpilibj2.command.SequentialCommandGroup;
```
Comment (tal-sitton) [inline src/main/java/frc/robot/OI.java:7]: why?

File: src/main/java/frc/robot/RobotMap.java (hunk 5 above)
Comment (tal-sitton) [inline src/main/java/frc/robot/RobotMap.java:25]: add `@todo`

File: src/main/java/frc/robot/RobotMap.java [hunk 10]
```diff
@@ -36,5 +39,11 @@
         int TRANSFER_ENTRANCE_LIGHT_SENSOR = 2;
 
         int INTAKE_TO_TRANSFER_LIMIT = 3;
+
+        // @TODO: change to real ports
+        int CLIMBER_PLACER_LEFT_LIMIT_FRONT = -1;
```
Comment (tal-sitton) [inline src/main/java/frc/robot/RobotMap.java:44]: check with mechatronics if all of those limits will be in the robot, maybe they forgot them

File: src/main/java/frc/robot/OI.java [hunk 11]
```diff
@@ -23,7 +24,8 @@
      * <b>Yellow Button</b>: Climber Up<br>
      * <b>Blue Button</b>: Climber Stop<br>
      * <b>D-Pad Down Button</b>: Reverse all<br>
-     * <b>D-Pad Left Button</b>: Reverse IntakeToTransfer and Intake
+     * <b>D-Pad Left Button</b>: Reverse IntakeToTransfer and Intake<br>
+     * <b>Xbox Left Joystick Button</b>: Move to next bar
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/OI.java:28]: That is still inconsistent

File: src/main/java/frc/robot/commands/MoveToNextBar.java [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+
+    private static final RootNamespace rootNamespace = new RootNamespace("move to next bar");
+
+    private static final Supplier<Double> RELEASE_HOOKS_TIMEOUT = rootNamespace.addConstantDouble("release hooks timeout", 1);
+    private static final Supplier<Double> MOVE_PLACERS_DOWN_TIMEOUT = rootNamespace.addConstantDouble("move placers down timeout", 1);
+
+    public MoveToNextBar() {
+        ClimberWinch climberWinch = ClimberWinch.getInstance();
+        addCommands(
+                new MoveGenericSubsystem(climberWinch, ClimberWinch.UP_SPEED).withTimeout(RELEASE_HOOKS_TIMEOUT.get()),
```
Comment (tal-sitton) [inline src/main/java/frc/robot/commands/MoveToNextBar.java:25]: why timeout?

File: src/main/java/frc/robot/commands/MoveToNextBar.java [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+                new ParallelCommandGroup(
```
Comment (tal-sitton) [inline src/main/java/frc/robot/commands/MoveToNextBar.java:26]: what's the purpose? add docs

File: src/main/java/frc/robot/commands/MoveToNextBar.java [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+import com.spikes2212.dashboard.RootNamespace;
+import edu.wpi.first.wpilibj2.command.ParallelCommandGroup;
+import edu.wpi.first.wpilibj2.command.SequentialCommandGroup;
+import frc.robot.subsystems.ClimberPlacer;
+import frc.robot.subsystems.ClimberWinch;
+
+import java.util.function.Supplier;
+
+/**
+ * Moves the dynamic climbers upwards until they hit the next bar.
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/commands/MoveToNextBar.java:13]: Fix documentation

File: src/main/java/frc/robot/commands/DropPlacer.java [hunk 15]
```diff
@@ -0,0 +1,37 @@
+package frc.robot.commands;
+
+import com.spikes2212.command.genericsubsystem.commands.MoveGenericSubsystem;
+import com.spikes2212.dashboard.RootNamespace;
+import edu.wpi.first.wpilibj.Timer;
+import frc.robot.subsystems.ClimberPlacer;
+
+import java.util.function.Supplier;
+
+public class DropPlacer extends MoveGenericSubsystem {
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/commands/DropPlacer.java:10]: Rename class (shouldn't be drop anymore)

File: src/main/java/frc/robot/subsystems/ClimberPlacer.java [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+ * Controls the position of the climber.
+ */
+public class ClimberPlacer extends MotoredGenericSubsystem {
+
+    public static final double MIN_SPEED = -0.6;
+    public static final double MAX_SPEED = 0.6;
+
+    private final Supplier<Double> stallCurrent = rootNamespace.addConstantDouble("stall current", 10);
+
+    public final Supplier<Double> DROP_SPEED = rootNamespace.addConstantDouble("drop speed", 0.5);
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/subsystems/ClimberPlacer.java:22]: Rename

Comment (tal-sitton) [inline src/main/java/frc/robot/subsystems/ClimberPlacer.java:22]: image

File: src/main/java/frc/robot/subsystems/ClimberPlacer.java [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+    public void configureDashboard() {
+        rootNamespace.putData("drop " + side + " placer", new DropPlacer(this));
+        rootNamespace.putData("drop both placers", new DropBothPlacers());
+    }
+
+    public boolean isStalling() {
+        return Math.abs(talon.getStatorCurrent()) > stallCurrent.get();
+    }
+
+    public WPI_TalonSRX getTalon() {
```
Comment (tal-sitton) [inline src/main/java/frc/robot/subsystems/ClimberPlacer.java:70]: when will u use it?

File: src/main/java/frc/robot/commands/DropBothPlacers.java [hunk 18]
```diff
@@ -0,0 +1,16 @@
+package frc.robot.commands;
+
+import edu.wpi.first.wpilibj2.command.ParallelCommandGroup;
+import frc.robot.subsystems.ClimberPlacer;
+
+public class DropBothPlacers extends ParallelCommandGroup {
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/commands/DropBothPlacers.java:6]: You are using this very simple `ParallelCommandGroup` only once, does it really deserve a class?

File: src/main/java/frc/robot/subsystems/ClimberPlacer.java [hunk 19]
```diff
@@ -0,0 +1,81 @@
...
+        this.backLimit = new DigitalInput(backLimitPort);
+    }
+
+    @Override
+    public boolean canMove(double speed) {
+        return !(frontLimit.get() && speed < 0) && !(backLimit.get() && speed > 0);
+    }
+
+    @Override
+    public void configureDashboard() {
```
Comment (tal-sitton) [inline src/main/java/frc/robot/subsystems/ClimberPlacer.java:61]: put `statorCurrent` in the dashboard

**Review by tal-sitton — CHANGES_REQUESTED** [review]

**Review by OfriRosenbaum — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/subsystems/ClimberPlacer.java (hunk 19 above)
Comment (yutytuty) [reply src/main/java/frc/robot/subsystems/ClimberPlacer.java:61]: It is in the dashboard. Line 20.

File: src/main/java/frc/robot/OI.java (hunk 11 above)
Comment (yutytuty) [reply src/main/java/frc/robot/OI.java:28]: How should it be then

Comment (OfriRosenbaum) [reply src/main/java/frc/robot/OI.java:28]: Move To Next Bar

Comment (OfriRosenbaum) [reply src/main/java/frc/robot/OI.java:28]: Actually, change all of the other to be like this

File: src/main/java/frc/robot/commands/DropBothPlacers.java (hunk 18 above)
Comment (yutytuty) [reply src/main/java/frc/robot/commands/DropBothPlacers.java:6]: Apparently yes. Ask Riki about single responsibility or some shit like that.

File: src/main/java/frc/robot/subsystems/ClimberPlacer.java (hunk 19 above)
Comment (tal-sitton) [reply src/main/java/frc/robot/subsystems/ClimberPlacer.java:61]: that's the desired `statorCurrent`, not the current `statorCurrent`

**Review by tal-sitton — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/RobotMap.java (hunk 10 above)
Comment (tal-sitton) [reply src/main/java/frc/robot/RobotMap.java:45]: ok, there is no limit switches for the placers for this district.

File: src/main/java/frc/robot/commands/MoveToNextBar.java [hunk 20]
```diff
@@ -0,0 +1,37 @@
...
+import com.spikes2212.dashboard.RootNamespace;
+import edu.wpi.first.wpilibj2.command.ParallelCommandGroup;
+import edu.wpi.first.wpilibj2.command.SequentialCommandGroup;
+import frc.robot.subsystems.ClimberPlacer;
+import frc.robot.subsystems.ClimberWinch;
+
+import java.util.function.Supplier;
+
+/**
+ * Moves the winch up a little to release the hooks. Then the command moves the hooks down towards the ground. Then the
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/commands/MoveToNextBar.java]: image

File: src/main/java/frc/robot/commands/MoveBothPlacersToNextBar.java [hunk 21]
```diff
@@ -0,0 +1,16 @@
+package frc.robot.commands;
+
+import edu.wpi.first.wpilibj2.command.ParallelCommandGroup;
+import frc.robot.subsystems.ClimberPlacer;
+
+public class MoveBothPlacersToNextBar extends ParallelCommandGroup {
+
+    public MoveBothPlacersToNextBar() {
+        ClimberPlacer leftPlacer = ClimberPlacer.getLeftInstance();
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/commands/MoveBothPlacersToNextBar.java]: just use the `getInstance()` on the command and call supper

File: src/main/java/frc/robot/commands/MoveBothPlacersToNextBar.java [hunk 22]
```diff
@@ -0,0 +1,16 @@
...
+        ClimberPlacer rightPlacer = ClimberPlacer.getRightInstance();
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/commands/MoveBothPlacersToNextBar.java]: image

File: src/main/java/frc/robot/commands/MovePlacerToNextBar.java [hunk 23]
```diff
@@ -0,0 +1,31 @@
...
+import com.spikes2212.command.genericsubsystem.commands.MoveGenericSubsystem;
+import com.spikes2212.dashboard.RootNamespace;
+import edu.wpi.first.wpilibj2.command.SequentialCommandGroup;
+import frc.robot.subsystems.ClimberPlacer;
+
+import java.util.function.Supplier;
+
+public class MovePlacerToNextBar extends SequentialCommandGroup {
+
+    private static final RootNamespace rootNamespace = new RootNamespace("move placers to next bar");
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/commands/MovePlacerToNextBar.java:12]: placers??? why plural

**Review by OfriRosenbaum — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/commands/MoveToNextBar.java (hunk 20 above)
Comment (yutytuty) [reply src/main/java/frc/robot/commands/MoveToNextBar.java]: image

File: src/main/java/frc/robot/Robot.java [hunk 24]
```diff
@@ -8,6 +8,7 @@
 import com.spikes2212.dashboard.RootNamespace;
 import edu.wpi.first.wpilibj.TimedRobot;
 import edu.wpi.first.wpilibj2.command.CommandScheduler;
+import frc.robot.subsystems.ClimberPlacer;
```
Comment (tal-sitton) [inline src/main/java/frc/robot/Robot.java]: why?

File: src/main/java/frc/robot/commands/MoveToNextBar.java [hunk 25]
```diff
@@ -0,0 +1,36 @@
...
+ * Must start the command <b>ONLY</b> when you are already on a bar.
+ */
+public class MoveToNextBar extends SequentialCommandGroup {
+
+    private static final RootNamespace rootNamespace = new RootNamespace("move to next bar");
+
+    private static final Supplier<Double> RELEASE_HOOKS_TIMEOUT = rootNamespace.addConstantDouble("release hooks timeout", 1);
+    private static final Supplier<Double> RAISE_PLACER_TIMEOUT = rootNamespace.addConstantDouble("raise placer timeout", 1);
+
+    public MoveToNextBar() {
```
Comment (tal-sitton) [inline src/main/java/frc/robot/commands/MoveToNextBar.java:23]: add docs for **every step in the way**

**Review by tal-sitton — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/Robot.java [hunk 26]
```diff
@@ -39,20 +40,26 @@ public void robotInit() {
...
+        rightClimberPlacer.configureDashboard();
 
         rootNamespace = new RootNamespace("robot namespace");
         rootNamespace.putData("intake cargo", new IntakeCargo());
         rootNamespace.putData("release cargo", new ReleaseCargo());
         rootNamespace.putData("drive forward", new DriveArcade(drivetrain, 0.5, 0));
         rootNamespace.putData("drive backward", new DriveArcade(drivetrain, -0.5, 0));
         rootNamespace.putData("move to cargo", new MoveToCargo(drivetrain, MoveToCargo.CARGO_MOVE_VALUE));
+        rootNamespace.putData("drop both placers", new MoveBothPlacersToNextBar());
+
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/Robot.java:62]: remove blank line

File: src/main/java/frc/robot/commands/MoveToNextBar.java [hunk 27]
```diff
@@ -0,0 +1,42 @@
...
+import edu.wpi.first.wpilibj2.command.SequentialCommandGroup;
+import frc.robot.subsystems.ClimberPlacer;
+import frc.robot.subsystems.ClimberWinch;
+
+import java.util.function.Supplier;
+
+/**
+ * Moves the robot to the next bar. <br>
+ * Steps: <br>
+ * 1. Move the winch up to release the hooks from the bar.<br>
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/commands/MoveToNextBar.java:15]: present simple

**Review by OfriRosenbaum — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/Robot.java (hunk 26 above)
Comment (yutytuty) [reply src/main/java/frc/robot/Robot.java:62]: Where? I do not see a blank line. There seems to have been a misunderstanding. After you submitted the review. I went to the exact location that is specified and I saw no blank line. only the line:         intakePlacer.setDefaultCommand(new IntakePlacerDefaultCommand());.

File: src/main/java/frc/robot/Robot.java [hunk 28]
```diff
@@ -9,7 +9,7 @@
 import edu.wpi.first.wpilibj.TimedRobot;
 import edu.wpi.first.wpilibj2.command.CommandScheduler;
 import frc.robot.commands.*;
-import frc.robot.commands.autonomous.*;
+import frc.robot.commands.autonomous.GyroAutonomous;
```
Comment (tal-sitton) [inline src/main/java/frc/robot/Robot.java:12]: why?

**Review by tal-sitton — CHANGES_REQUESTED** [review]

**Pull request merged**